### PR TITLE
semantic_analyser: validate use of calls as map keys

### PR DESCRIPTION
**Pull request by mmarchini**
Fixes: https://github.com/iovisor/bpftrace/issues/697
Fixes: https://github.com/iovisor/bpftrace/issues/633

### CODE REVIEW

File: src/ast/semantic_analyser.cpp [hunk 1]
```diff
@@ -1259,16 +1259,40 @@ bool SemanticAnalyser::is_final_pass() const
   return pass_ == num_passes_;
 }
 
-bool SemanticAnalyser::check_assignment(const Call &call, bool want_map, bool want_var)
+bool SemanticAnalyser::check_assignment(const Call &call, bool want_map, bool want_var, bool want_map_key)
 {
-  if (want_map && want_var)
+  if (want_map && want_var && want_map_key)
```
Comment (danobi) [inline src/ast/semantic_analyser.cpp]: Is is possible to want an assignment to a map, variable, _and_ map key?

File: src/ast/semantic_analyser.cpp [hunk 2]
```diff
@@ -1285,11 +1309,19 @@ bool SemanticAnalyser::check_assignment(const Call &call, bool want_map, bool wa
       return false;
     }
   }
+  else if (want_map_key)
+  {
+    if (!call.key_for_map)
+    {
+      err_ << call.func << "() should be assigned to a variable" << std::endl;
```
Comment (fbs) [inline src/ast/semantic_analyser.cpp]: s/variable/map key/  right?

File: src/ast/semantic_analyser.cpp (hunk 1 above)
Comment (danobi) [reply src/ast/semantic_analyser.cpp]: Nevermind, realizing now `check_assignment(call, true, true, true)` reads as "it's ok to assign me to a map/variable or use me as a map key".

File: src/ast/semantic_analyser.cpp (hunk 2 above)
Comment (danobi) [reply src/ast/semantic_analyser.cpp]: That might read weird. Maybe
```suggestion
      err_ << call.func << "() should be used as a map key" << std::endl;
```

**Comment by danobi** [comment]
@mmarchini  ping

**Comment by danobi** [comment]
Merged in #975

**Pull request closed without merging**